src/System.Management.Automation/engine/remoting/fanin/InitialSessionStateProvider.cs

```diff
@@ -2890,4 +2927,79 @@ internal static string[] TryGetStringArray(object hashObj)
         }
     }
     #endregion
+
+    #region DISCFileValidation
+
+    internal static class DISCFileValidation
+    {
+        // Set of supported configuration options for a PowerShell InitialSessionState.
+        private static readonly HashSet<string> SupportedConfigOptions = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase)
+        {
+            "SchemaVersion",
+            "GUID",
+            "Author",
+            "Description",
+            "CompanyName",
+            "Copyright",
+            "SessionType",
+            "TranscriptDirectory",
+            "MountUserDrive",
+            "UserDriveMaximumSize",
+            "ScriptsToProcess",
+            "LanguageMode",
+            "ExecutionPolicy",
+            "ModulesToImport",
+            "VisibleAliases",
+            "VisibleCmdlets",
+            "VisibleFunctions",
+            "VisibleExternalCommands",
+            "VisibleProviders",
+            "AliasDefinitions",
+            "FunctionDefinitions",
+            "VariableDefinitions",
+            "EnvironmentVariables",
+            "TypesToProcess",
+            "FormatsToProcess",
+            "AssembliesToLoad"
+        };
+
+        // These are configuration options for WSMan (WinRM) endpoint configurations, that 
+        // appearand in .pssc files, but are not part of PowerShell InitialSessionState.
+        private static readonly HashSet<string> UnsupportedConfigOptions = new HashSet<string>(StringComparer.InvariantCultureIgnoreCase)
+        {
+            "RunAsVirtualAccount",
+            "RunAsVirtualAccountGroups",
+            "GroupManagedServiceAccount",
+            "RoleDefinitions",
+            "RequiredGroups",
+            "PowerShellVersion"
+        };
+
+        internal static void ValidateContents(Hashtable configHash)
+        {
+            foreach (var key in configHash.Keys)
+            {
+                if (!(key is string keyName))
+                {
+                    throw new PSInvalidOperationException(RemotingErrorIdStrings.DISCInvalidConfigKeyType);
+                }
+
+                if (!SupportedConfigOptions.Contains(keyName))
+                {
+                    if (UnsupportedConfigOptions.Contains(keyName))
+                    {
+                        throw new PSInvalidOperationException(
+                            StringUtil.Format(RemotingErrorIdStrings.DISCUnsupportedConfigName, keyName));
+                    }
```

Comment:
If we throw on "unsupported" why do we nest this in `if (!SupportedConfigOptions.Contains(keyName))` condition? Should it has higher priority?